pkg/auth/providers/githubapp/githubapp_provider.go

```diff
@@ -0,0 +1,480 @@
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/rancher/pkg/user"
+	wcorev1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	// Name is used to reference the Provider.
+	Name = "githubapp"
+
+	userType = "user"
+	teamType = "team"
+	orgType  = "org"
+)
+
+type githubClient interface {
+	getAccessToken(ctx context.Context, code string, config *cattlev3.GithubAppConfig) (string, error)
+	getUser(ctx context.Context, githubAccessToken string, config *cattlev3.GithubAppConfig) (Account, error)
+	getOrgsForUser(ctx context.Context, username string, config *cattlev3.GithubAppConfig) ([]Account, error)
+	getTeamsForUser(ctx context.Context, username string, config *cattlev3.GithubAppConfig) ([]Account, error)
+	searchUsers(ctx context.Context, searchTerm, searchType string, config *cattlev3.GithubAppConfig) ([]Account, error)
+	searchTeams(ctx context.Context, searchTerm string, config *cattlev3.GithubAppConfig) ([]Account, error)
+	getUserOrgByID(ctx context.Context, id int, config *cattlev3.GithubAppConfig) (Account, error)
+	getTeamByID(ctx context.Context, id int, config *cattlev3.GithubAppConfig) (Account, error)
+}
+
+type tokensManager interface {
+	GetSecret(userID string, provider string, fallbackTokens []accessor.TokenAccessor) (string, error)
+	IsMemberOf(token accessor.TokenAccessor, group v3.Principal) bool
+	CreateTokenAndSetCookie(userID string, userPrincipal v3.Principal, groupPrincipals []v3.Principal, providerToken string, ttl int, description string, request *types.APIContext) error
+	UserAttributeCreateOrUpdate(userID, provider string, groupPrincipals []v3.Principal, userExtraInfo map[string][]string, loginTime ...time.Time) error
+}
+
+type userManager interface {
+	CheckAccess(accessMode string, allowedPrincipalIDs []string, userPrincipalID string, groups []v3.Principal) (bool, error)
+	SetPrincipalOnCurrentUser(apiContext *types.APIContext, principal v3.Principal) (*v3.User, error)
+}
+
+type ghAppProvider struct {
+	ctx           context.Context
+	authConfigs   v3.AuthConfigInterface
+	secrets       wcorev1.SecretController
+	getConfig     func() (*cattlev3.GithubAppConfig, error)
+	githubClient  githubClient
+	userManager   userManager
+	tokenMGR      tokensManager
+	appDataLoader func(ctx context.Context, appID int64, privateKey []byte, installationID int64, endpoint string) (*gitHubAppData, error)
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userManager user.Manager, tokenMGR *tokens.Manager) common.AuthProvider {
+	provider := &ghAppProvider{
+		ctx:          ctx,
+		authConfigs:  mgmtCtx.Management.AuthConfigs(""),
+		secrets:      mgmtCtx.Wrangler.Core.Secret(),
+		githubClient: &githubAppClient{httpClient: http.DefaultClient},
```

Comment:
I think we should not use the default http client but a custom one with a reasonable (whatever that is) timeout set
```suggestion
		githubClient: &githubAppClient{httpClient: &http.Client{Timeout: 30 * time.Second}},
```